CONTRIBUTING.md

```diff
@@ -0,0 +1,60 @@
+
+In sum: if you are interested in integrating your code into core OTP, a significant amount of the time you invest in OTP will need to be spent on collaboration, discussion, documentation, etc. and you will need to be available for regular meetings. You will need to take this into consideration in your budget and timeline.
+
+## Other ways to share development work
+
+We don't want to discourage innovation and experimentation, and want promising new features to be visible to other users, so we have also created a Sandbox system for fast-track review and inclusion in mainline OTP. Most code for a Sandbox feature must be located in its own package, and any code added to the core of OTP must be in conditional blocks controlled by a feature flag and disabled by default. This greatly simplifies and accelerates the review process, as requirements on the code outside core OTP will be much less stringent. Please see http://docs.opentripplanner.org/en/latest/SandboxExtension/ for more details.
+
+If you don't have the time to participate in this process, the community is of course still interested in the new ideas and capabilities in your fork of OTP. Please do share them on the groups (mailing lists) where you may create awareness of your work and attract collaborators with more resources. The goal of mainline OTP is not to be everything to everyone, but rather to contain the most solid code relied upon daily by the primary OTP contributor organizations (as well as Sandbox features that have been cleanly isolated from the core system). So in a sense it's encouraged for people to work on special-purpose forks.
+
+## Additional Considerations for Structuring Pull Requests
+
+When creating and building on a pull request, please do the following:
+
+- Describe in detail the problem you are solving and any major design choices you have made. PR descriptions must clearly state how your changes work.
+- Break large changes down into a series of smaller logical steps in separate PRs.
+- Tie such series of PRs together with an "epic issue" that explains the overall plan.
+- Use Github issue references ("addresses #12" etc.) to connect PRs and issues together.
+- Consider squashing and rebasing to make the PR history easier to understand, eliminating extra "noise" commits like accidental changes to organization specific code, abandoned experiments, or reverted configuration changes.
+- If possible, please describe your proposed changes in advance before proceeding too far with development. There are often other people working on the same problem. Coordinating with them can avoid a lot of duplicated effort, conflict, and technical debt.
```

Comment:
I personally think this is the most important thing - so maybe we should put it first. Should we write _discuss_ instead of _describe_:

```suggestion
- If possible, please discuss your proposed changes in advance before proceeding too far with development. There are often other input/angles/thoughts and/or plans on the same problem. Coordinating with other OTP developers can avoid a lot of duplicated effort, conflict, and technical debt and make the review a lot easier.
```
> There are often other input/angles/thoughts and/or plans on the same problem.